test/addons_test.go

```diff
@@ -214,6 +214,61 @@ func TestAzureGroup(t *testing.T) {
 // Private Functions
 // -----------------------------------------------------------------------------
 
+func checkIfUpgradeIsNeeded(t *testing.T, groupname string) (bool, []v1beta2.AddonInterface, error) {
+	t.Logf("determining which addons in group %s need to be upgrade tested", groupname)
+
+	var doUpgrade bool
+	addons := groups[groupname]
+	for _, addon := range addons {
+		if err := overrides(addon); err != nil {
+			return false, nil, err
+		}
+	}
+
+	addonDeploymentsArray := []v1beta2.AddonInterface{}
+	for _, newAddon := range addons {
+		t.Logf("verifying whether upgrade testing is needed for addon %s", newAddon.GetName())
+		oldAddon, err := addontesters.GetLatestAddonRevisionFromLocalRepoBranch("../", comRepoRemote, comRepoRef, newAddon.GetName())
+		if err != nil {
+			if strings.Contains(err.Error(), "directory not found") {
+				t.Logf("no need to upgrade test %s, it appears to be a new addon (no previous revisions found in branch %s)", newAddon.GetName(), comRepoRef)
+				addonDeploymentsArray = append(addonDeploymentsArray, newAddon)
+				continue
+			}
+			return false, nil, err
+		}
+		if oldAddon == nil {
+			t.Logf("no need to upgrade test %s, it appears to be a new addon (no previous revisions found in branch %s)", newAddon.GetName(), comRepoRef)
+			addonDeploymentsArray = append(addonDeploymentsArray, newAddon)
+			continue // new addon, upgrade test not needed
+		}
+
+		t.Logf("determining old and new versions for upgrade testing addon %s", newAddon.GetName())
+		oldRev := oldAddon.GetAnnotations()[constants.AddonRevisionAnnotation]
+		oldVersion, err := semver.Parse(strings.TrimPrefix(oldRev, "v"))
+		if err != nil {
+			return false, nil, err
+		}
+		newRev := newAddon.GetAnnotations()[constants.AddonRevisionAnnotation]
+		newVersion, err := semver.Parse(strings.TrimPrefix(newRev, "v"))
+		if err != nil {
+			return false, nil, err
+		}
+
+		if newVersion.EQ(oldVersion) {
+			t.Logf("skipping upgrade test for addon %s, it has not been updated", newAddon.GetName())
+			addonDeploymentsArray = append(addonDeploymentsArray, oldAddon)
+			continue
+		} else if oldVersion.GT(newVersion) {
+			return false, nil, fmt.Errorf("revisions for addon %s are broken, previous revision %s is newer than current %s", newAddon.GetName(), oldVersion, newVersion)
+		}
```

Comment:
i wonder why we need to do this comparison, shouldn't the controller handle it....

Reply:
It is for the sake of creating addonUpgrades objects, which has little to do how kubernetes handles resources.